lib/pos_consensus_controller.go

```diff
@@ -62,13 +65,82 @@ func (cc *ConsensusController) HandleFastHostStuffVote(event *consensus.FastHotS
 	// 4. Broadcast the timeout msg to the network
 }
 
-func (cc *ConsensusController) HandleFastHostStuffTimeout(event *consensus.FastHotStuffEvent) {
-	// The consensus module has signaled that we have timed out for a view. We construct and
-	// broadcast the timeout here:
-	// 1. Verify the block height and view we want to timeout on are valid
-	// 2. Construct the timeout message
-	// 3. Process the timeout in the consensus module
-	// 4. Broadcast the timeout msg to the network
+// HandleFastHostStuffTimeout is triggered when the FastHotStuffEventLoop has signaled that
+// it is ready to time out the current view. This function validates the timeout signal for
+// staleness. If the signal is valid, then it construct and broadcast the timeout msg here.
+//
+// Steps:
+// 1. Verify the timeout message's and the view we want to timeout on
+// 2. Construct the timeout message
+// 3. Process the timeout in the consensus module
+// 4. Broadcast the timeout msg to the network
+func (cc *ConsensusController) HandleFastHostStuffTimeout(event *consensus.FastHotStuffEvent) error {
+	// Hold a read lock on the consensus controller. This is because we need to check the
+	// current view and block height of the consensus module.
+	cc.lock.Lock()
+	defer cc.lock.Unlock()
+
+	var err error
+
+	if !consensus.IsProperlyFormedTimeoutEvent(event) {
+		// If the event is not properly formed, we ignore it and log it. This should never happen.
+		return errors.Errorf("HandleFastHostStuffTimeout: Received improperly formed timeout event: %v", event)
+	}
+
+	if event.View != cc.fastHotStuffEventLoop.GetCurrentView() {
+		// It's possible that the event loop signaled to timeout, but at the same time, we
+		// received a block proposal from the network and advanced the view. This is normal
+		// and an expected race condition in the steady-state.
+		//
+		// Nothing to do here.
+		return errors.Errorf("HandleFastHostStuffTimeout: Stale timeout event: %v", event)
+	}
+
+	// Locally advance the event loop's view so that the node is locally running the Fast-HotStuff
+	// protocol correctly. Any errors below related to broadcasting the timeout message should not
+	// affect the correctness of the protocol's local execution.
+	if _, err := cc.fastHotStuffEventLoop.AdvanceViewOnTimeout(); err != nil {
+		// This should never happen as long as the event loop is running. If it happens, we return
+		// the error and let the caller handle it.
+		return errors.Errorf("HandleFastHostStuffTimeout: Error advancing view on timeout: %v", err)
+	}
+
+	// Construct the timeout message
+	timeoutMsg := NewMessage(MsgTypeValidatorTimeout).(*MsgDeSoValidatorTimeout)
+	timeoutMsg.MsgVersion = MsgValidatorTimeoutVersion0
+	timeoutMsg.TimedOutView = event.View
+	// TODO: Somehow we need to know the validator's ECDSA public key. Fill this out once the
+	// mapping between BLS and ECDSA keys is implemented and available in the consensus module.
+	// timeoutMsg.PublicKey = <ECDSA public key here>
```

Comment:
@AeonSw4n, @lazynina we need the validator's ECDSA public key available in this controller for timeout and vote construction. We'll similarly need it for block construction. A few of options in how we can do it:
- We pull it from the on-chain state by constructing a UtxoView at the current tip to fetch the current validator set's ValidatorEntries
- We maintain a cached version of the validator set in this controller for each block height. We need to compute the validator set for each safe block anyway when processing each incoming block
- The validator's ECDSA public key is configured along with the BLS signer metadata and passed along into the controller

My preference would be option 2. LMK if you have other thoughts

Reply:
Sounds good. I'll follow up with another PR for it